nexus/networking/src/firewall_rules.rs

```diff
@@ -439,3 +486,100 @@ pub async fn plumb_service_firewall_rules(
     .await?;
     Ok(())
 }
+
+/// Compute firewall rules implied by an IP allowlist.
+///
+/// This looks up the source IP allowlist in the database, and generates
+/// firewall rules attached to each provided NIC that implements the allowlist.
+/// If the allowlist is "any", there are no restrictions and no rules are
+/// returned.
+///
+/// Otherwise, two sets of rules are returned:
+///
+/// - A low-priority rule that denies all traffic
+/// - A higher-priority rule that allows it from the specific IPs.
+async fn allowlist_to_firewall_rules(
+    nics: &[NetworkInterface],
+    allowed_ips: Vec<IpNet>,
+) -> Option<Vec<sled_agent_client::types::VpcFirewallRule>> {
+    if nics.is_empty() {
+        return None;
+    }
+
+    // Convert to host identifiers, how the sled-agent types expose these.
+    assert!(!allowed_ips.is_empty(), "Allowlist must not be empty here");
+    let filter_hosts = Some(
+        allowed_ips
+            .into_iter()
+            .map(|net| sled_agent_client::types::HostIdentifier::Ip(net.into()))
+            .collect(),
+    );
+
+    let low_priority_deny = sled_agent_client::types::VpcFirewallRule {
+        status: VpcFirewallRuleStatus::Enabled.into(),
+        direction: VpcFirewallRuleDirection::Inbound.into(),
+        targets: nics.to_vec(),
+        filter_hosts: filter_hosts.clone(),
+        filter_ports: Some(vec![L4PortRange {
+            first: 1.try_into().unwrap(),
+            last: u16::MAX.try_into().unwrap(),
+        }
+        .into()]),
+        filter_protocols: None,
+        action: VpcFirewallRuleAction::Deny.into(),
+        priority: ALLOW_LIST_LOW_PRIORTY,
+    };
+
+    let allow = sled_agent_client::types::VpcFirewallRule {
+        status: VpcFirewallRuleStatus::Enabled.into(),
+        direction: VpcFirewallRuleDirection::Inbound.into(),
+        targets: nics.to_vec(),
+        filter_hosts,
+
+        // TODO-correctness: Having no filter on ports or protocols is not
+        // techniically correct. Nexus should only receive TCP traffic, on HTTP
+        // or HTTPs traffic, for example. However, at this point in the system,
+        // we don't really have access to _which service_ this NIC is for, only
+        // that it is for _one_ of our services.
+        //
+        // We may want to reorganize this code to apply these rules on a
+        // per-service basis. This is discussed in RFD 474.
```

Comment:
With apologies for the naivety of this question - will this allow IPs on the allowlist to connect to ports within the zone that we don't want to expose? I'm thinking of something like external DNS, which runs both DNS (should allow connections) and a dropshot server (should only allow connections from within the rack).

Reply:
I think it would, though I thought those listened on different addresses. Isn't the Dropshot server on the underlay, while the DNS server listens through an OPTE port? I will check the code now.

Reply:
Yeah, I think you're right. I just hopped on a dogfood external DNS zone, and it's only listening on TCP on the underlay, and only on UDP on the explicit OPTE address.